src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs

```diff
@@ -47,34 +50,67 @@ private class OnAssemblyLoadInitializer
         private readonly ILifespanManager _lifespanManager;
         private readonly string _requiredAssemblyName;
 
+#if NETFRAMEWORK
+        private int _firstInitialization = 1;
+#endif
+
         public OnAssemblyLoadInitializer(ILifespanManager lifespanManager, InstrumentationInitializer instrumentationInitializer)
         {
             _instrumentationInitializer = instrumentationInitializer;
             _lifespanManager = lifespanManager;
             _requiredAssemblyName = instrumentationInitializer.RequiredAssemblyName;
+
             AppDomain.CurrentDomain.AssemblyLoad += CurrentDomain_AssemblyLoad;
+
+#if NETFRAMEWORK
+            var isRequiredAssemblyLoaded = AppDomain.CurrentDomain
+                .GetAssemblies()
+                .Any(x => GetAssemblyName(x) == _requiredAssemblyName);
+            if (isRequiredAssemblyLoaded)
+            {
+                OnRequiredAssemblyDetected();
+            }
+#endif
         }
 
         private void CurrentDomain_AssemblyLoad(object sender, AssemblyLoadEventArgs args)
         {
-            var assemblyName = args.LoadedAssembly.FullName.Split(new[] { ',' }, count: 2)[0];
+            var assemblyName = GetAssemblyName(args.LoadedAssembly);
 
             if (_requiredAssemblyName == assemblyName)
             {
-                var initializerName = _instrumentationInitializer.GetType().Name;
-                Logger.Debug("'{0}' started", initializerName);
-
-                try
-                {
-                    _instrumentationInitializer.Initialize(_lifespanManager);
-                }
-                catch (Exception ex)
-                {
-                    Logger.Error(ex, "'{0}' failed", initializerName);
-                }
-
-                AppDomain.CurrentDomain.AssemblyLoad -= CurrentDomain_AssemblyLoad;
+                OnRequiredAssemblyDetected();
+            }
+        }
+
+        private void OnRequiredAssemblyDetected()
+        {
+#if NETFRAMEWORK
+            if (Interlocked.Exchange(ref _firstInitialization, value: 0) != 1)
```

Comment:
Why this is needed?

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/1605#discussion_r1019553855 third bullet.